Expose the SslStream and Cipher Suite on ITlsHandshakeFeature

amcasey opened the pull request:
# Expose the SslStream and Cipher Suite on ITlsHandshakeFeature

<!-- If this is your first PR in the ASP.NET Core repo, please run through the checklist
below to ensure a smooth review and merge process for your PR. -->

- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

<!-- Once all that is done, you're ready to go. Open the PR with the content below. -->

As was sensibly requested in #46442, we should expose these properties - their omission was likely an oversight.

Fixes #46442
Fixes #46823

amcasey committed:
Expose TlsCipherSuite on ITlsHandshakeFeature

...in .NET 7 and newer.  Return `null` by default to avoid breaking existing implementations.

Part of #46442

amcasey committed:
Implement ITlsHandshakeFeature.NegotiatedCipherSuite in TlsConnectionFeature

amcasey committed:
Expose SslStream on ITlsHandshakeFeature

...in .NET 7 and newer.  Return `null` by default to avoid breaking existing implementations.

Part of #46442

amcasey committed:
Implement ITlsHandshakeFeature.SslStream in TlsConnectionFeature

Tratcher commented on src/Servers/Connections.Abstractions/src/Features/ITlsHandshakeFeature.cs:
```diff
@@ -15,6 +19,18 @@ public interface ITlsHandshakeFeature
+    /// <summary>
+    /// Gets the <see cref="SslStream"/>.
+    /// </summary>
+    SslStream? SslStream { get => default; }
```
SslStream will only ever work on Kestrel, it should be on a separate feature. It's at least theoretically possible for the other servers to implement NegotiatedCipherSuite from the data they have.

amcasey replied:
Do I call it `ISslStreamFeature` or `ITlsSslStreamFeature`? 🙃

Tratcher commented on src/Servers/Kestrel/Core/src/Internal/TlsConnectionFeature.cs:
```diff
@@ -65,7 +66,21 @@ public SslProtocols Protocol
 
+    public SslStream? SslStream
+    {
+        get => _sslStream;
```
Another part of #46442 is making sure _sslStream is consistently set by all of the UseHttps APIs. Right now only some of them set this field.

amcasey replied:
Oops, I knew that and I blanked.

amcasey replied:
Okay, several hops later...

`TlsConnectionFeature` always has an `SslStream`, but `TlsConnectionFeature` is only constructed if `HttpsConnectionMiddleware.OnConnectionAsync` is called.  This suggests some `UseHttps` overloads are not missing merely the `SslStream`, but all of `ITlsConnectionFeature`, `ITlsApplicationProtocolFeature`, and `ITlsHandshakeFeature`.  Or I'm missing something...

amcasey replied:
Also, assuming my comment from my previous change is correct, even the overloads of `UseHttps` that reference `OnConnectionAsync` won't actually invoke it if only http/3 is enabled.

amcasey replied:
Staring at the code, I can't find an overload that doesn't ultimately use `HttpsConnectionMiddleware.OnConnectionAsync` (assuming http/1 and/or http/2 is enabled).  Do you have a scenario in mind @Tratcher?

Tratcher replied:
Ah, my bad. It's not _sslStream that's null, it's that we don't consistently set the SslStream into the feature collection. 
https://github.com/dotnet/aspnetcore/blob/7d0b688e7b2241290346a98901708ea574db7f69/src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs#L419
Your new feature addresses that problem. I'd suggest removing this line that adds SslStream directly to the feature collection, but that would probably break people.

Tratcher replied:
We still don't support ITlsConnectionFeature for HTTP/3? I remember asking MsQuic to expose that information, but I don't remember if we ever got it.

Tratcher replied:
Nevermind: https://github.com/dotnet/runtime/issues/70184

amcasey committed:
Move SslStream into a kestrel-specific feature

halter73 commented on src/Servers/Kestrel/Core/src/Internal/TlsConnectionFeature.cs:
```diff
@@ -11,13 +11,14 @@
 namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal;
 
-internal sealed class TlsConnectionFeature : ITlsConnectionFeature, ITlsApplicationProtocolFeature, ITlsHandshakeFeature
+internal sealed class TlsConnectionFeature : ITlsConnectionFeature, ITlsApplicationProtocolFeature, ITlsHandshakeFeature, ISslStreamFeature
```
We still need to explicitly set this in the middleware. It won't automatically be retrievable as an `ISslStreamFeature` just because it now implements the interface.

https://github.com/dotnet/aspnetcore/blob/33e7a9a111742c8b06924383baba5c448d7f8af8/src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs#L143-L149

We should add an ISslStreamFeature test to HttpsConnectionMiddlewareTests and update HandshakeDetailsAreAvailable to include NegotiatedCipherSuite.

https://github.com/dotnet/aspnetcore/blob/33e7a9a111742c8b06924383baba5c448d7f8af8/src/Servers/Kestrel/test/InMemory.FunctionalTests/HttpsConnectionMiddlewareTests.cs#L103-L111

amcasey replied:
> We still need to explicitly set this in the middleware. It won't automatically be retrievable as an ISslStreamFeature just because it now implements the interface.

I had no idea.  Is that an encapsulation thing, so people don't tug on interfaces that haven't been explicitly made accessible?

amcasey replied:
> We should add an ISslStreamFeature test to HttpsConnectionMiddlewareTests and update HandshakeDetailsAreAvailable to include NegotiatedCipherSuite.

I'd be interested to hear more about the test philosophy here.  Intuitively, this is not something I'd expect to be able to break (once it's working in the first place), so it's not something I'd naturally devote a lot of test execution time to.

amcasey replied:
Done, but I'd still like to know more about how we decide what needs test coverage.

amcasey commented on src/Servers/Connections.Abstractions/src/Features/ITlsHandshakeFeature.cs:
```diff
@@ -15,6 +19,13 @@ public interface ITlsHandshakeFeature
+    /// <summary>
+    /// Gets the <see cref="TlsCipherSuite"/>.
+    /// </summary>
+    TlsCipherSuite? NegotiatedCipherSuite { get => default; }
```
Do we ever use `default` in this codebase?

amcasey committed:
Prefer `null` to `default`

Co-authored-by: Stephen Halter <halter73@gmail.com>

amcasey committed:
Explicitly add the ISslStreamFeature to the collection

amcasey committed:
Add test coverage

JamesNK commented on src/Servers/Kestrel/Core/src/Features/ISslStreamFeature.cs:
```diff
@@ -0,0 +1,17 @@
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Features;
+
+/// <summary>
+/// Feature to set access the TLS application protocol
```
```suggestion
/// Feature to get access the connection's <see cref="SslStream" />.
```

JamesNK commented on src/Servers/Kestrel/Core/src/Features/ISslStreamFeature.cs:
```diff
@@ -0,0 +1,17 @@
+    /// <summary>
+    /// Gets the <see cref="SslStream"/>.
+    /// </summary>
+    SslStream SslStream { get; }
```
An http connection returns a null `ISslStreamFeature`? Therefore SslStream is never null?

amcasey replied:
I think you're asking why there's no nullability annotation.  This interface property is just exposing a property that's already on the concrete type, where it's non-nullable.  There's a comment thread above with Chris about when it's actually available, but as far as I've been able to determine, the feature(s) are present if and only if `UseHttps` is called.

Tratcher replied:
Right, the feature should only be added if SslStream is available. E.g. don't add it for HTTP/3 or non-TLS.

amcasey replied:
I think that's how it behaves now, but I wanted to confirm that you meant "should" in the normative sense and not in the "you missed something" sense.

Tratcher replied:
The code looks correct. I hadn't reviewed it fully when I commented on this.

amcasey committed:
Correct copy-paste error

Co-authored-by: James Newton-King <james@newtonking.com>

Tratcher commented on src/Servers/Kestrel/Core/src/Features/ISslStreamFeature.cs:
```diff
@@ -0,0 +1,17 @@
+public interface ISslStreamFeature
+{
+    /// <summary>
+    /// Gets the <see cref="SslStream"/>.
```
```suggestion
    /// Gets the <see cref="SslStream"/>.
    /// This is not available for non-TLS connections or HTTP/3.
```

amcasey replied:
I'm not sure I follow - seems like either the comment should be on the interface or the property should be optional?

Tratcher replied:
It does belong on the interface, but I put it here because people are more likely to see it via intellasense. You could put it on both.

Tratcher commented on src/Servers/Connections.Abstractions/src/Features/ITlsHandshakeFeature.cs:
```diff
@@ -15,6 +19,13 @@ public interface ITlsHandshakeFeature
+    /// <summary>
+    /// Gets the <see cref="TlsCipherSuite"/>.
+    /// </summary>
+    TlsCipherSuite? NegotiatedCipherSuite { get => null; }
```
In theory we could assembly this from the other properties with a big lookup table, but the maintenance on that wouldn't be great and you'd always have some nulls for unknown combinations. I guess we can try that later.

amcasey replied:
I considered that and concluded that it would be really, really painful.

Tratcher commented on src/Servers/Kestrel/Core/src/Internal/TlsConnectionFeature.cs:
```diff
@@ -65,7 +66,21 @@ public SslProtocols Protocol
     // We don't store the values for these because they could be changed by a renegotiation.
+
+    TlsCipherSuite? ITlsHandshakeFeature.NegotiatedCipherSuite { get => NegotiatedCipherSuite; }
+    // Split the writable property from the interface implementation because it doesn't make sense to allow setting null
+    internal TlsCipherSuite NegotiatedCipherSuite
+    {
+        get => _cipherSuite ?? _sslStream.NegotiatedCipherSuite;
+        set => _cipherSuite = value;
+    }
```
Why is it settable? It doesn't seem to be available anywhere. It seems like you don't need this property at all, you could directly return _sslStream.NegotiatedCipherSuite.

Actually, only ClientCertificate is directly settable from the interface. None of these others need setters.

amcasey replied:
Just trying to follow local conventions.  Always happy to delete setters.  Do you want me to clean up all of them?

amcasey commented on src/Servers/Connections.Abstractions/src/Features/ITlsHandshakeFeature.cs:
```diff
@@ -3,6 +3,10 @@
 
 using System.Security.Authentication;
 
+#if NET7_0_OR_GREATER
```
That was copy-pasted.  Do you want me to update `TlsConnectionCallbackContext` and `TlsConnectionCallbackOptions` too?

Tratcher replied:
Yes please

BrennanConroy replied:
I actually like using the versioned defines, it's self-documenting for when the feature was added.
And this one would actually be NET8_0_OR_GREATER.

amcasey replied:
I'm getting some mixed messages here. 😆

amcasey replied:
On the one hand, @BrennanConroy is actually on the team that owns this code.  On the other hand, that seems like a really flaky way to capture introduction versions, especially since the type has no actual dependency on .net 8.  I'm leaning `NETCOREAPP` unless someone feels strongly.

Tratcher replied:
Git blame is better for code history 😁 
NET8_0 implies there's something specific required from .NET 8 that wasn't available previously. I'll check if I can find any other examples, but I think we've used the more general defines elsewhere.

Tratcher replied:
We're inconsistent at best 😆 
https://github.com/dotnet/aspnetcore/search?q=NETCOREAPP
https://github.com/dotnet/aspnetcore/search?q=NET7_0_OR_GREATER

I'd still suggest the most general option.

BrennanConroy replied:
We kind of do look consistent actually. We use versioned defines for things that actually require the version or higher, and non-versioned for things that need more than netstandard2.0/2.1. Which actually goes against what I originally said here. DIMs work in netcoreapp so that's probably the define that should be used.

amcasey committed:
Add comments about when ISslStreamFeature is expected to be available

amcasey committed:
Use more accurate ifdefs

amcasey committed:
Drop unused setters

amcasey commented:
Force push is a trivial rebase.  API Review changes to follow.

amcasey committed:
Simplify declaration of ITlsHandshakeFeature.NegotiatedCipherSuite

amcasey committed:
Set SslStream feature consistently

Directly setting a concrete type we don't control as a feature is an anti-pattern, but we were already doing it some scenarios, so we should at least do it predictably.  Set it if and only if the `ISslStreamFeature` will be added.

BrennanConroy commented on src/Servers/Kestrel/Core/src/Features/ISslStreamFeature.cs:
```diff
@@ -0,0 +1,19 @@
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Features;
+
+/// <summary>
+/// Feature to get access the connection's <see cref="SslStream" />.
```
```suggestion
/// Feature to get access to the connection's <see cref="SslStream" />.
```

amcasey committed:
Fix doc typo

Co-authored-by: Brennan <brecon@microsoft.com>
